confluent_kafka/schema_registry/json_schema.py

```diff
@@ -0,0 +1,261 @@
+
+    .. _Schema definition:
+        https://json-schema.org/understanding-json-schema/reference/generic.html
+
+    """  # noqa: E501
+    __slots__ = ['_hash', '_auto_register', '_known_subjects', '_parsed_schema',
+                 '_registry', '_schema', '_schema_id', '_schema_name',
+                 '_subject_name_func', 'to_dict']
+
+    # default configuration
+    _default_conf = {'auto.register.schemas': True,
+                     'subject.name.strategy': topic_subject_name_strategy}
+
+    def __init__(self, schema_registry_client, schema_str, to_dict=None,
+                 conf=None):
+        self._registry = schema_registry_client
+        self._schema_id = None
+        # Avoid calling registry if schema is known to be registered
+        self._known_subjects = set()
+
+        if to_dict is not None and not callable(to_dict):
+            raise ValueError("to_dict must be callable with the signature"
+                             " to_dict(Serialization Context, object)->dict")
+
+        self.to_dict = to_dict
+
+        # handle configuration
+        conf_copy = self._default_conf.copy()
+        if conf is not None:
+            conf_copy.update(conf)
+
+        self._auto_register = conf_copy.pop('auto.register.schemas')
+        if not isinstance(self._auto_register, bool):
+            raise ValueError("auto.register.schemas must be a boolean value")
+
+        self._subject_name_func = conf_copy.pop('subject.name.strategy')
+        if not callable(self._subject_name_func):
+            raise ValueError("subject.name.strategy must be callable")
+
+        if len(conf_copy) > 0:
+            raise ValueError("Unrecognized properties: {}"
+                             .format(conf_copy.keys()))
+
+        schema_dict = json.loads(schema_str)
+        schema_name = schema_dict.get('title', None)
+        if schema_name is None:
+            raise ValueError("Missing required JSON schema annotation Title")
+
+        self._schema_name = schema_name
+        self._parsed_schema = schema_dict
+        self._schema = Schema(schema_str, schema_type="JSON")
+
+    def __call__(self, ctx, obj):
+        """
+        Serializes an object to the Confluent Schema Registry's JSON binary
+        format.
+
+        Args:
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+            obj (object): object instance to serializes.
+
+
+        Note:
+            None objects are represented as Kafka Null.
+
+        Raises:
+            SerializerError if any error occurs serializing obj
+
+        Returns:
+            bytes: Confluent Schema Registry formatted JSON bytes
+
+        """
+        if obj is None:
+            return None
+
+        subject = self._subject_name_func(ctx, self._schema_name)
+
+        # Check to ensure this schema has been registered under subject_name.
+        if self._auto_register and subject not in self._known_subjects:
```

Comment:
i think locking of _known_subjects is a good idea because if one call to SR is underway you probably don't want another one as well (since it will almost certainly be for the same schema).

Reply:
I thought about preventing multiple inflight requests but I decided to forego locking as it makes things easier to reason about. This does mean you will initially have a few more requests than intended but I'd argue its hardly a problem. If you feel strong about it we could add locking but since we will always get the same schema id (the schema is fixed) I didn't feel it was necessary.

Reply:
I'm with Ryan, we discussed this for Avro and decided that the extra calls to  SR are okay at this time.

Reply:
Resolving for now, we can tweak it later if we feel this is too burdensome.

Reply:
i still disagree, weighing up the magnitude of the potential downsides, but i think in practice worst case scenario unlikely to be hit, so i don't really care.